File: misc/rotate-array[2].c
void rotate(int nums[], int n, int k) {
    int i = 0;
    int count = 0;
    int start = 0;
    int cur = 0;
    int tmp = nums[0];

    if(k <= 0)
        return ;


    while(count < n){ 
        do{ 
            nums[start] = nums[(k+cur)%n];
            nums[(k+cur)%n] = tmp;
            tmp = nums[start];
            cur = (k+cur) % n;
            count++;
        }while(cur != start);

        start++;
        cur = start;
        tmp = nums[cur];
    }   
}

File: go/array/rotate-array[2].go
func rotate(nums []int, k int)  {
    k %= len(nums)

    reverse(nums, 0, len(nums) - k - 1)
    reverse(nums, len(nums) - k, len(nums) - 1)
    reverse(nums, 0, len(nums) - 1)
}

func reverse(nums []int, start, end int){
    for(start < end){
        nums[start], nums[end] = nums[end], nums[start]
        start++
        end--
    }
}
